src/main/java/de/florianmichael/viafabricplus/injection/mixin/fixes/minecraft/screen/hud/MixinPlayerListHud.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * This file is part of ViaFabricPlus - https://github.com/FlorianMichael/ViaFabricPlus
+ * Copyright (C) 2021-2024 FlorianMichael/EnZaXD <florian.michael07@gmail.com> and RK_01/RaphiMC
+ * Copyright (C) 2023-2024 contributors
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package de.florianmichael.viafabricplus.injection.mixin.fixes.minecraft.screen.hud;
+
+import de.florianmichael.viafabricplus.injection.access.IPlayerListEntry;
+import de.florianmichael.viafabricplus.injection.access.IPlayerListHud;
+import de.florianmichael.viafabricplus.settings.impl.VisualSettings;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.client.gui.hud.PlayerListHud;
+import net.minecraft.client.network.PlayerListEntry;
+import net.minecraft.network.ClientConnection;
+import org.spongepowered.asm.mixin.Final;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.Shadow;
+import org.spongepowered.asm.mixin.Unique;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Mixin(PlayerListHud.class)
+public abstract class MixinPlayerListHud implements IPlayerListHud {
+
+    @Unique
+    private static final Comparator<PlayerListEntry> viaFabricPlus$FIFO_COMPARATOR
+            = Comparator.comparingInt(ple -> ((IPlayerListEntry) ple).viaFabricPlus$getIndex());
+
+    @Shadow
+    @Final
+    private MinecraftClient client;
+
+    @Unique
+    private int viaFabricPlus$maxSlots = 60;
```

Comment:
Where do you get this constant from? Looking at the 1.7 source code they just use a normal ArrayList for tablist entries (which should lead into the fifo sorting you already implemented) But I don't get where/why this should be done.

Reply:
The vanilla client in 1.7 times, set the size of the tablist to use whatever the server sent was the [max online players in the login packet](https://wiki.vg/index.php?title=Protocol&oldid=6003#Join_Game). If you set a vanilla server to max players = 60, you'd get a 3 column and 20 row tablist. If you set a vanilla server to max players = 200, you'd get a 10 column 20 row tablist which had tiny squares and was completely unreadable.

Bukkit's (or at least many forks) server-side limited the max players field to 60 [to avoid the list going unreadable](https://bugs.mojang.com/browse/MC-61676), you can find [30 examples](https://github.com/search?q=playerconnection.sendPacket%28new+PacketPlayOutLogin%28entityplayer.getId%28%29%2C+entityplayer.playerInteractManager.getGameMode%28%29%2C+worlddata.isHardcore%28%29%2C+worldserver.worldProvider.getDimension%28%29%2C+worldserver.getDifficulty%28%29%2C+Math.min%28this.getMaxPlayers%28%29%2C+60%29%2C+worlddata.getType%28%29%2C+worldserver.getGameRules%28%29.getBoolean%28%22reducedDebugInfo%22%29%29%29%3B+%2F%2F+CraftBukkit+-+cap+player+list+to+60&type=code) on github like [this one](https://github.com/CobbleSword/BurritoSpigot/blob/cb667b112c6b8b4f9af9f8403b6fed5d649a419f/BurritoSpigot-Server/nms-patches/PlayerList.patch#L117) that comes alongside a comment `// CraftBukkit - cap player list to 60`

What the implementation of the mod does, is set these `max slots` field to whatever comes in the login packet, the 60 is just an ancient default that was very common to see due to all that is explained above, and should do as a fallback shall everything fail (a future version where max players isn't sent in the login packet? idk)

Reply:
Thanks for your very detailed explaination,

I would prefer to have either no value at all or a value that results in the same behavior as in the 1.7 client if you remove the part of the code that replaces the field with the value sent by the server (of course it makes no difference here because the default value is never used). My point here is simply that we don't have any constants in the code that don't come from Minecraft or can't be logically derived otherwise so that when you look at the code later you won't be confused.